### PR TITLE
cmake: Use INTERFACE libraries to expose seqan3 to users

### PR DESCRIPTION
**Pull request by marehr**
This cmake changes makes it more easy to include seqan3, it also makes the following macros obsolete

* `SEQAN3_INCLUDE_DIRS`
* `SEQAN3_LIBRARIES`
* `SEQAN3_DEFINITIONS`
* `SEQAN3_CXX_FLAGS`

```cmake
add_executable (myapp myapp.cpp)
target_link_libraries (myapp seqan::seqan3)

# is the same as explicitly setting: 
# SEQAN3_INCLUDE_DIRS, SEQAN3_LIBRARIES, SEQAN3_DEFINITIONS, SEQAN3_CXX_FLAGS
```

### CODE REVIEW

File: build_system/seqan3-config.cmake [hunk 1]
```diff
@@ -599,6 +599,15 @@ if (NOT ${FIND_NAME}_FIND_QUIETLY)
     message (STATUS "${ColourBold}Found SeqAn3:${ColourReset} ${SEQAN3_BASEDIR}/seqan3 (found version \"${SEQAN3_VERSION_STRING}\")")
 endif ()
 
+separate_arguments (SEQAN3_CXX_FLAGS_LIST UNIX_COMMAND "${SEQAN3_CXX_FLAGS}")
```
Comment (h-2) [inline build_system/seqan3-config.cmake]: this splits on ' '?

File: build_system/seqan3-config.cmake [hunk 2]
```diff
@@ -599,6 +599,15 @@ if (NOT ${FIND_NAME}_FIND_QUIETLY)
...
+string(REGEX REPLACE ";( )*[-/]D" ";" SEQAN3_DEFINITIONS_LIST ";${SEQAN3_DEFINITIONS};")
```
Comment (h-2) [inline build_system/seqan3-config.cmake]: this splits on other special characters, as well?

Do wo know that the above work work for all the flags possibly set by seqan3-config?

File: build_system/seqan3-config.cmake [hunk 3]
```diff
@@ -599,6 +599,15 @@ if (NOT ${FIND_NAME}_FIND_QUIETLY)
...
+
+add_library (seqan::seqan3 INTERFACE IMPORTED)
```
Comment (h-2) [inline build_system/seqan3-config.cmake]: why did you call it "seqan::seqan3" and not just "seqan2?

File: test/header/CMakeLists.txt [hunk 4]
```diff
@@ -77,12 +76,8 @@ macro (seqan3_header_test component header_base_path)
     file (WRITE "${PROJECT_BINARY_DIR}/header_test_src/empty.cpp" "")
 
     add_executable (${target} ${PROJECT_BINARY_DIR}/header_test_src/empty.cpp)
-    target_compile_options (${target} PRIVATE ${cxx_flags_list} ${SEQAN3_TEST_CXX_FLAGS})
-    target_compile_definitions (${target} PRIVATE ${SEQAN3_DEFINITIONS} ${SEQAN3_TEST_DEFINITIONS})
-    target_include_directories (${target} PRIVATE ${SEQAN3_INCLUDE_DIRS} ${SEQAN3_TEST_INCLUDE_DIRS})
-    target_link_libraries (${target} ${SEQAN3_LIBRARIES} ${SEQAN3_TEST_LIBRARIES})
+    target_link_libraries (${target} seqan3_unit_test)
```
Comment (h-2) [inline test/header/CMakeLists.txt]: where do we actually tell cmake to use the previously defined "library"?

File: test/header/CMakeLists.txt [hunk 5]
```diff
@@ -102,17 +97,18 @@ macro (seqan3_header_test component header_base_path)
 TEST(Header, ${header_source_test_name}) {}")
 
             add_library (${target_library} OBJECT "${source_file}")
-            target_compile_options (${target_library} PRIVATE ${cxx_flags_list} ${SEQAN3_TEST_CXX_FLAGS})
-            target_compile_definitions (${target_library} PRIVATE ${SEQAN3_DEFINITIONS} ${SEQAN3_TEST_DEFINITIONS})
-            target_include_directories (${target_library} PRIVATE ${SEQAN3_INCLUDE_DIRS} ${SEQAN3_TEST_INCLUDE_DIRS})
+            # NOTE: a simple target_link_libraries (${target_library} seqan3_unit_test)
+            # is not possible for OBJECT libraries
```
Comment (h-2) [inline test/header/CMakeLists.txt]: whats an "object library"?

File: test/seqan3-test.cmake [hunk 6]
```diff
@@ -48,22 +48,25 @@ include (FindPackageMessage)
...
 set (SEQAN3_TEST_CLONE_DIR "${PROJECT_BINARY_DIR}/vendor/googletest")
 
-# required flags, includes, definitions and libraries for seqan3
-set (SEQAN3_STRICT_CXX_FLAGS "-pedantic" "-Werror" "-Wall" "-Wextra")
+# required flags, includes, definitions and libraries for seqan3 tests
+add_library (seqan3_test INTERFACE)
+target_link_libraries (seqan3_test INTERFACE "seqan::seqan3")
+target_link_libraries (seqan3_test INTERFACE "pthread")
+target_compile_options (seqan3_test INTERFACE "-pedantic" "-Werror" "-Wall" "-Wextra")
+target_include_directories (seqan3_test INTERFACE "${SEQAN3_CLONE_DIR}/test/include/")
```
Comment (h-2) [inline test/seqan3-test.cmake]: Did I understand this correctly:
we define a new "library interface" that depends on another interface "seqan::seqan3", but also has additional properties?

File: test/seqan3-test.cmake [hunk 7]
```diff
@@ -117,9 +120,7 @@ macro (seqan3_require_benchmark)
     add_library (benchmark STATIC IMPORTED)
     add_dependencies (benchmark google_benchmark)
     set_target_properties (benchmark PROPERTIES IMPORTED_LOCATION "${benchmark_library}")
-
-    list (APPEND SEQAN3_BENCHMARK_LIBRARIES "benchmark")
-    list (APPEND SEQAN3_BENCHMARK_LIBRARIES "pthread")
+    set_property (TARGET benchmark APPEND PROPERTY INTERFACE_LINK_LIBRARIES "pthread")
```
Comment (h-2) [inline test/seqan3-test.cmake]: Why is this required? We have:
`target_link_libraries (seqan3_test INTERFACE "pthread")`
and
`target_link_libraries (seqan3_performance_test INTERFACE "benchmark")`

File: test/seqan3-test.cmake [hunk 8]
```diff
@@ -162,10 +163,7 @@ macro (seqan3_require_test)
     add_library (gtest STATIC IMPORTED)
     add_dependencies (gtest gtest_main)
     set_target_properties (gtest PROPERTIES IMPORTED_LOCATION "${gtest_library}")
-
-    list (APPEND SEQAN3_TEST_LIBRARIES "gtest_main")
-    list (APPEND SEQAN3_TEST_LIBRARIES "gtest")
-    list (APPEND SEQAN3_TEST_LIBRARIES "pthread")
+    set_property (TARGET gtest APPEND PROPERTY INTERFACE_LINK_LIBRARIES "pthread")
```
Comment (h-2) [inline test/seqan3-test.cmake]: here again

File: test/seqan3-test.cmake [hunk 9]
```diff
@@ -162,10 +163,7 @@ macro (seqan3_require_test)
...
 endmacro ()
 
 macro (add_subdirectories)
```
Comment (h-2) [inline test/seqan3-test.cmake]: I am still a little confused about the entities in this file and naming. Is "benchmark" actually a "dependency handler" for "google_benchmark"? Could we rename it gbenchmark, like for "gtest"?

Could we also rename  our new "interface names" to be consistent with the filesystem hierarchy, i.e.
* `seqan3` instead of seqan::seqan3
* `seqan3_test` (stays the same)
* `seqan3_test_unit` instead of seqan3_unit_test
* `seqan3_test_performance` instead of seqan3_performance_test

Comment (h-2) [reply test/seqan3-test.cmake]: And should seqan3_test_unit and seqan3_test_performance not be moved to their respective folders' CMakeLists.txt?

File: build_system/seqan3-config.cmake (hunk 1 above)
Comment (marehr) [reply build_system/seqan3-config.cmake]: yes; whitespace

File: build_system/seqan3-config.cmake (hunk 2 above)
Comment (marehr) [reply build_system/seqan3-config.cmake]: Yes, since DEFINITIONS should only define `-DMY_DEF` `/DMY_DEF`. This is inspired by `FindMPI`. If you want to add compiler flags you should use `target_compile_options`.

If we would require cmake 3.11 we could use [target_compile_definitions](https://cmake.org/cmake/help/latest/command/target_compile_definitions.html) directly. Before, it was not allowed to use the `target_*` functions to add properties to a `INTERFACE IMPORTED`

File: build_system/seqan3-config.cmake (hunk 3 above)
Comment (marehr) [reply build_system/seqan3-config.cmake]: it's cool to have a namespace we could later have 
```cmake
add_library (seqan3::seqan3 INTERFACE IMPORTED)
add_library (seqan3::zlib INTERFACE IMPORTED)
add_library (seqan3::bzip INTERFACE IMPORTED)
```

And the user could pull the changes in.

Comment (marehr) [reply build_system/seqan3-config.cmake]: I would suggest the name change to `seqan3::seqan3`

File: test/header/CMakeLists.txt (hunk 5 above)
Comment (marehr) [reply test/header/CMakeLists.txt]: It is a pseudo-library. If you add sources to `add_executable(target a.cpp b.cpp)` when internally you will get for each `.cpp` file a `.o` file. The problem is if you change any header/source file you have to rebuild every `.cpp` file of the `target`. What I do here is to build the `.o` for each `.cpp` individually. Such that a source change will only affect a subset off all `.cpp` files.

File: test/seqan3-test.cmake (hunk 6 above)
Comment (marehr) [reply test/seqan3-test.cmake]: Yep, it's transitive. https://cmake.org/cmake/help/v3.0/manual/cmake-buildsystem.7.html#transitive-usage-requirements

File: test/seqan3-test.cmake (hunk 7 above)
Comment (marehr) [reply test/seqan3-test.cmake]: google benchmark itself needs pthread. If you don't specify it here it will be at the wrong position in the positions of the flags and the benchmark library might not know some symbols.

We can drop `target_link_libraries (seqan3_test INTERFACE "pthread")`, because it is at the wrong place/position.

File: test/seqan3-test.cmake (hunk 9 above)
Comment (marehr) [reply test/seqan3-test.cmake]: > I am still a little confused about the entities in this file and naming. Is "benchmark" actually a "dependency handler" for "google_benchmark"? Could we rename it gbenchmark, like for "gtest"?
Yep good idea!

> Could we also rename  our new "interface names" to be consistent with the filesystem hierarchy, i.e.
> * `seqan3` instead of seqan::seqan3
> * `seqan3_test` (stays the same)
> * `seqan3_test_unit` instead of seqan3_unit_test
> * `seqan3_test_performance` instead of seqan3_performance_test

Good idea, I would stay with `seqan3::seqan3`, since every recent cmake-module which exposes a 'interface library' does it through a namespaced one. For example Mpi exposes `mpi::mpi`.

File: test/header/CMakeLists.txt (hunk 4 above)
Comment (marehr) [reply test/header/CMakeLists.txt]: `seqan3_unit_test` inherits from `seqan3::seqan3`

**Comment by marehr** [comment]
> Also, I think its confusing that we now set things like CXXFLAGS via target_link_libraries. Is there no generic way of saying "get all the things from the interface"?

This is the generic way ;) We link against a header only library xD

>  Also if we add this, please document this in the top of `seqan3-config.cmake` where the other exported variables are also documented.

Will do :+1:

File: build_system/seqan3-config.cmake (hunk 3 above)
Comment (h-2) [reply build_system/seqan3-config.cmake]: What's the point of of seqan3::zlib? Wouldnt that be part of seqan3 anyway? :thinking:

File: test/header/CMakeLists.txt (hunk 4 above)
Comment (h-2) [reply test/header/CMakeLists.txt]: yep I saw that later on

**Comment by h-2** [comment]
I think we agree on all points (how unusual :yum: )!

**Comment by h-2** [comment]
@marehr Can you address the last points so we can merge this?

**Comment by marehr** [comment]
@h-2 please review again

File: test/unit/test/tmp_filename_test.cpp [hunk 10]
```diff
@@ -34,8 +34,10 @@
 
 #include <gtest/gtest.h>
 #include <fstream>
-#include "tmp_filename.hpp"
 
+#include <seqan3/test/tmp_filename.hpp>
+
+namespace fs = seqan3::filesystem;
```
Comment (h-2) [inline test/unit/test/tmp_filename_test.cpp]: why was this added here?

**Comment by h-2** [comment]
> Also if we add this, please document this in the top of seqan3-config.cmake where the other exported variables are also documented.

I tihnk you have not addressed this comment of mine, yet.

File: test/seqan3-test.cmake [hunk 11]
```diff
@@ -48,22 +48,31 @@ include (FindPackageMessage)
...
-set (SEQAN3_TEST_INCLUDE_DIRS "")
-set (SEQAN3_TEST_LIBRARIES "")
-list (APPEND SEQAN3_TEST_INCLUDE_DIRS "${SEQAN3_TEST_CLONE_DIR}/googletest/include/")
-list (APPEND SEQAN3_TEST_INCLUDE_DIRS "${SEQAN3_CLONE_DIR}/test/include/")
+# seqan3::test exposes a base set of required flags, includes, definitions and
+# libraries qhich are in common for **all** seqan3 tests
+add_library (seqan3::test INTERFACE IMPORTED)
+set_property (TARGET seqan3::test APPEND PROPERTY INTERFACE_COMPILE_OPTIONS "-pedantic"  "-Wall" "-Wextra" "-Werror")
+set_property (TARGET seqan3::test APPEND PROPERTY INTERFACE_LINK_LIBRARIES "seqan3::seqan3" "pthread")
+set_property (TARGET seqan3::test APPEND PROPERTY INTERFACE_INCLUDE_DIRECTORIES "${SEQAN3_CLONE_DIR}/test/include/")
```
Comment (rrahn) [inline test/seqan3-test.cmake]: Here again, I am not sure, why these library targets need to be defined in the top level config cmake.
I'd like to have it as clean as possible in order to avoid the mess we had before. 
Is there any particular reason, of why these targets are exposed here and not within the project they are needed?
As far as I could see, there is no cross linking between other projects happening.

File: test/unit/test/tmp_filename_test.cpp (hunk 10 above)
Comment (marehr) [reply test/unit/test/tmp_filename_test.cpp]: the test case didn't work and I fixed it.

The test case was originally in `/test/include/seqan3/test/tmp_filename_test.cpp` (in the header file folder for test stuff) where it was never build, executed and tested.

I moved it to `/test/unit/test/tmp_filename_test.cpp` and all of the sudden this never tested file needed some changes.

File: test/seqan3-test.cmake (hunk 11 above)
Comment (marehr) [reply test/seqan3-test.cmake]: This is the top-level-config for all test cases, not the top-level-config of `find(seqan3)`.

And `seqan3::test` is the parent (base class) to all kinds of tests (e.g. header test, unit test, performance test, coverage test), so it needs to be defined in the top-level.

There are also some library definition like `seqan3::test::unit`, which is used as a base for a couple of other kinds of test cases. For example `seqan3::test::coverage`, `seqan3::test::header` inherit from it. Thus, this needs to be defined in the top-level.

Comment (rrahn) [reply test/seqan3-test.cmake]: >This is the top-level-config for all test cases, not the top-level-config of find(seqan3).

I know. But independent of this.

>And seqan3::test is the parent (base class) to all kinds of tests (e.g. header test, unit test, performance test, coverage test), so it needs to be defined in the top-level.
>There are also some library definition like seqan3::test::unit, which is used as a base for a couple of other kinds of test cases. For example seqan3::test::coverage, seqan3::test::header inherit from it. Thus, this needs to be defined in the top-level.

Ok, I just see now the dependencies. Alternatively, one could add a .cmake file into each subproject, but I see that the overhead might not be worthwhile. Could you add a section header above this definitions? Calling it `target link libraries for test projects` or something.

**Comment by marehr** [comment]
!rebuild

**Comment by marehr** [comment]
!rebuild
